src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/TypeNameParser.NativeAot.cs

```diff
@@ -113,7 +117,7 @@ private bool CheckTopLevelAssemblyQualifiedName()
             Justification = "GetType APIs are marked as RequiresUnreferencedCode.")]
         [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern",
             Justification = "GetType APIs are marked as RequiresUnreferencedCode.")]
-        private Type? GetType(string typeName, ReadOnlySpan<string> nestedTypeNames, string? assemblyNameIfAny)
+        private Type? GetType(string escapedTypeName, ReadOnlySpan<string> nestedTypeNames, AssemblyName? assemblyNameIfAny, string _)
```

Comment:
Why did this flip to pass escaped names around? 

It would make more sense to me pass unescaped names around and only do the escaping on the very edges where it is needed for backward compatibility.

Reply:
My main goal was to make `TypeName.Name|FullName|AssemblyQualifiedName` return always the same values as the corresponding APIs in `Type` (so the entry barrier is lower as the users don't need to learn new concepts and/or subtle API differences). And all the `Type` properties always return escaped values.

Since some of the CLR and NativeAOT methods expect escaped names and all Mono methods expect escaped names, I decided to switch to escaped by default. Also, unescaping is faster than escaping (this is what some code paths had to do before) as it's searching for one specific character rather than few that need the escaping.

I am very soon going to share perf numbers so hopefully I can convince you.

BTW the whole escaping had 0 test coverage before https://github.com/dotnet/runtime/pull/98562

Reply:
I think we are missing an API in the design.

If somebody wants to build a type resolver that resolves the type from metadata - like the ILLinker and AOT compilers, the escaped name is useless for that. They will need an unescaped name to resolve the type from metadata. We should have an API for that.

Reply:
What would the behavior of APIs like AssemblyQualifiedName name or FullName on constructed types be in this mode?

Would they throw (easiest option) or would they return unescaped names that are potentially invalid?

Reply:
> Would they throw (easiest option) or would they return unescaped names that are potentially invalid?

I was thinking about just returning the unescaped names. Because this is what these tools need?

The only alternative idea that I currently have it to expose following method:

```
class TypeName
{
+    public static string Unescape(string name);
}
```

Because these tools may need both `Name` and `FullName` in the unescaped forms (Name for nested types and full name for all other types).

Reply:
Note that my comment was about constructed types (instantiated generic types, arrays, pointers, byrefs).

> I was thinking about just returning the unescaped names. Because this is what these tools need?

Unescaped FullName or AssemblyQualifiedName would be ambiguous for constructed types, that makes it useless for any tools.

Reply:
I am not sure if I understand.

Let's consider two following examples:
1. A simple type with escaped name: `NotANested\+Type`
2. A pointer to the same type: `NotANested\+Type*`

What would be the expected behavior (and why) for the tools that need unescaped names?

Reply:
If the tool asks for a unescaped name on a pointer, it is a bug in the tool. The unescaped name would be ambiguous - it can mean either pointer or a simple type name ending with a star.

If we were to add TypeNameParseOptions.EscapeNames, we can prevent this pit of failure by throwing InvalidOperation exception instead of returning the ambiguous name in this case.

> public static string Unescape(string name)

This would address the scenario too. It is API review question whether it is preferable to handle this scenario via the `TypeName.Unscape` method or via ` TypeNameParseOptions.EscapeNames` option.

Reply:
> If the tool asks for a unescaped name on a pointer, it is a bug in the tool

Please correct me if I am wrong:
- for simple types (not an array, pointer, byref or generic), these tools need unescaped names
- for non-simple types the tools should never ask for the names, as they always resolve generic type definition or element type (simple types) and then use methods like `type.MakePointerType()`

Example:

```cs
TypeName parsed = TypeName.Parse(input);

if (parsed.IsPointer)
{
    TypeName elementTypeName = parsed.GetElementType();
    Type elementType = Resolve(elementTypeName.FullName);
    return elementType.MakePointerType();
}
```

Reply:
@jkotas I've spent more time looking at escaping and unescaping. Here are my findings:

1. Mono needs escaped input.
2. Tools need unescaped input (BTW we have no test coverage for that).
3. CLR and NativeAOT need both (escaped and unescaped).

Because CLR and NativeAOT need both, we can not introduce a property that says whether all names should be escaped or not.

When it comes to introducing a dedicated method for unescaping like `public static string Unescape(string name)` I am currently not convinced that we need public API for that.
All our parsers that need this method share code anyway and currently they simply use the internal implementation (30 lines of code). Because it's internal, we can make some assumptions like only valid names are provided (characters that should not be escaped are not escaped).
Morever since there is no spec that describes in detail what can be escaped and what can not, not making it public allows me to keep the status quo.

It would be great if you could review the PR so I could merge it before Preview 4 snap.

If you are going to insist on making this method part of the public API I hope that you don't mind if we keep it internal for now and release v1 of `TypeName` without it.

Reply:
> Mono needs escaped input.

It is because Mono still has a private copy of the parser in C/C++. If/once Mono is fully switched to the managed parser, it would be on the same plan as CLR and NativeAOT.

> CLR and NativeAOT need both (escaped and unescaped).

They need escaped names only for interop with other reflection APIs (like the Type.GetType with type resolve callback). They do not need the escaped names for the mainline `Type.GetType` on runtime type.

I think what you got here is fine, for now at least. I agree that the method to unescape the names is very simple.